### PR TITLE
Make keystore return `None` when a key doesn't exist

### PR DESCRIPTION
**Pull request by bkchr**
polkadot companion: https://github.com/paritytech/polkadot/pull/2492

### CODE REVIEW

File: client/authority-discovery/src/worker.rs [hunk 1]
```diff
@@ -299,7 +299,9 @@ where
 			// sign_with_all returns Result<Signature, Error> signature
 			// is generated for a public key that is supported.
 			// Verify that all signatures exist for all provided keys.
-			let signature = sign_result.map_err(|_| Error::MissingSignature(key.clone()))?;
+			let signature = sign_result.ok()
```
Comment (pepyakin) [inline client/authority-discovery/src/worker.rs:302]: Needs update of the comment

**Review by rakanalh — APPROVED** [review]
LGTM

File: client/keystore/src/local.rs [hunk 2]
```diff
@@ -411,36 +416,52 @@ impl KeystoreInner {
...
-			return Ok(phrase.clone())
+			return Ok(Some(phrase.clone()))
+		} else if self.path.is_none() {
+			return Ok(None)
 		}
 
-		let path = self.key_file_path(public, key_type).ok_or_else(|| Error::Unavailable)?;
-		let file = File::open(path)?;
+		let path = self.key_file_path(public, key_type)
+			.expect("We checked above that path exists; qed");
```
Comment (andresilva) [inline client/keystore/src/local.rs:427]: We could have removed the `self.path.is_none()` check above and just return `Ok(None)` here if `self.key_file_path` returns `None`.

File: client/keystore/src/local.rs [hunk 3]
```diff
@@ -278,24 +278,29 @@ impl SyncCryptoStore for LocalKeystore {
...
 	fn sr25519_vrf_sign(
 		&self,
 		key_type: KeyTypeId,
 		public: &Sr25519Public,
 		transcript_data: VRFTranscriptData,
-	) -> std::result::Result<VRFSignature, TraitError> {
+	) -> std::result::Result<Option<VRFSignature>, TraitError> {
 		let transcript = make_transcript(transcript_data);
 		let pair = self.0.read().key_pair_by_type::<Sr25519Pair>(public, key_type)
 			.map_err(|e| TraitError::PairNotFound(e.to_string()))?;
```
Comment (andresilva) [inline client/keystore/src/local.rs:293]: IMO this error no longer makes sense, if the pair wasn't there we'd be returning `None` from `key_pair_by_type` right?

File: primitives/keystore/src/lib.rs [hunk 4]
```diff
@@ -324,13 +328,13 @@ pub trait SyncCryptoStore: CryptoStore + Send + Sync {
 	/// each key given that the key is supported.
 	///
 	/// Returns a list of `Result`s each representing the SCALE encoded
-	/// signature of each key or a Error for non-supported keys.
+	/// signature of each key, `None` if the key doesn't exist or a error when something failed.
```
Comment (andresilva) [inline primitives/keystore/src/lib.rs:331]:
```suggestion
	/// signature of each key, `None` if the key doesn't exist or an error when something failed.
```

**Review by andresilva — CHANGES_REQUESTED** [review]
We should probably remove the `Error::PairNotFound` variant.

File: client/keystore/src/local.rs [hunk 5]
```diff
@@ -60,9 +60,9 @@ impl LocalKeystore {
 
 	/// Get a key pair for the given public key.
 	///
-	/// This function is only available for a local keystore. If your application plans to work with
-	/// remote keystores, you do not want to depend on it.
-	pub fn key_pair<Pair: AppPair>(&self, public: &<Pair as AppKey>::Public) -> Result<Pair> {
+	/// Returns `Ok(None)` if the key doesn't exist, `Ok(Some(_))` if the key exist or
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/keystore/src/local.rs:63]: This struct is called `LocalKeystore`. I think this makes it 100% clear that this isn't any remote keystore.

[the remark this answers is not in the document]
Comment (andresilva) [reply client/keystore/src/local.rs]: 🧠

File: client/keystore/src/local.rs (hunk 2 above)
Comment (bkchr) [reply client/keystore/src/local.rs:427]: :brain:

File: client/keystore/src/local.rs [hunk 6]
```diff
@@ -411,36 +415,49 @@ impl KeystoreInner {
...
-	fn key_phrase_by_type(&self, public: &[u8], key_type: KeyTypeId) -> Result<String> {
+	fn key_phrase_by_type(&self, public: &[u8], key_type: KeyTypeId) -> Result<Option<String>> {
 		if let Some(phrase) = self.get_additional_pair(public, key_type) {
-			return Ok(phrase.clone())
+			return Ok(Some(phrase.clone()))
 		}
 
-		let path = self.key_file_path(public, key_type).ok_or_else(|| Error::Unavailable)?;
-		let file = File::open(path)?;
+		let path = self.key_file_path(public, key_type)?;
```
Comment (andresilva) [inline client/keystore/src/local.rs:423]:
```suggestion
		let path = if let Some(path) = self.key_file_path(public, key_type) {
			path
		} else {
			return Ok(None);
		};
```

**Review by andresilva — CHANGES_REQUESTED** [review]
In the end we have to do almost the same thing as before 🙈

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Pull request merged**